### PR TITLE
Add support for updating sketches from an SD card

### PR DESCRIPTION
**Pull request by sandeepmistry**
Via 2nd stage boot code that checks SD card for `UPDATE.bin`.

Also, added tool info to `platform.txt`  to update sketches via HTTP outside the core for later.

cc/ @cmaglie @facchinm @agdl @akash73

### CODE REVIEW

**Comment by tuxedo0801** [comment]
Is there some kind of documentation on how to use this? Just copy the update.bin to sd-card and reboot arduino zero? Is this feature now part of the default-bootloader that comes with ArduinoCore-samd?

As SPI Flash is connected the same way as sd card, I guess it will work with SPI flash as well?

File: variants/arduino_zero/linker_scripts/gcc/flash_with_bootloader.ld [hunk 1]
```diff
@@ -67,6 +67,9 @@ SECTIONS
 	{
 		__text_start__ = .;
 
+		KEEP(*(.ota_boot))
+
+		. = ALIGN(256);
```
Comment (facchinm) [inline variants/arduino_zero/linker_scripts/gcc/flash_with_bootloader.ld]: I'd align the .ota_boot section to either 0x2000 or 0x4000 to make sure that the reset vector does not get relocated to something different than 0x6000 if the 2nd stage bootloader becomes slightly smaller

Comment (sandeepmistry) [inline variants/arduino_zero/linker_scripts/gcc/flash_with_bootloader.ld]: I'll change it to `0x2000`.

Maybe we can have a clearer name for this section `second_stage`, `second_stage_boot` or `sketch_boot`?

**Comment by tuxedo0801** [comment]
*ping*

**Comment by cmaglie** [comment]
@tuxedo0801 
you must patient a bit, this is still experimental, we'll provide documentation once ready.

**Comment by tuxedo0801** [comment]
@cmaglie thanks for first feedback.
I don't want the complete documentation. Just let me know if this would be tied to sd cards or if it will be possible with SPI flash as well (or maybe it's up to the user?!).

That would help me continuing my development into the right direction...

**Comment by cmaglie** [comment]
For now it's SD card only.

**Comment by dlabun** [comment]
Out of curiosity, is this going to replace the default boot loader or run in tandem with it?

**Comment by sandeepmistry** [comment]
@dlabun good question! It will run in tandem.

**Comment by tuxedo0801** [comment]
To give you an insight in why I want to use SPI flash instead of SD card:

I'm developing small devices for KNX home automation bus. The devices are totally bus-powered. And the KNX transceiver chip allows max ~50mA to be drawn from KNX bus. So it's very low power.

SD cards consume too much energy to be powered from KNX bus. SPI flash consumes almost nothing, compared to SD card. And even micro-sd card socket is much bigger than a small SPI flash package... Size matters ;-) 

As the SPI flash is not pluggable like SD card, the new sketch is received directly from KNX bus. So updating a device is just a simple click on the control-application on a PC that is also connected to bus. File is transferred over bus and stored into SPI flash.

I guess there are more people out there that wand to read the firmware from somewhere else than a pluggable SD card (get via ethernet, wifi, some radio module, power line, ... whatever and stored on any storage).

I would be glad if you foresee a user-friendly mechanism that allows the sketch-writer to speficy an own storage mechanism to read the update from, that don't requires fiddling with core classes and modifying the samd core.

Beside my whishes: Tumbs up for this new feature... Keep up the good work.

**Comment by deladriere** [comment]
what CS pin are you using ? (mine is pin 10)

I can't have the bootloader to boot from the SD
Here is what I did
1) installed as instructed
2) burned the bootloader with Atmel SAM-ICE programmer
3) compiled and renamed the .bin to UPDATE.bin
4) copied the UPDATE.bin to sd
5) reseted the board

but it stays in booloader mode with the led glowing 
Am i missing something ?

**Comment by deladriere** [comment]
Sorry to insist : I can't figure how to get started
thanks !

**Comment by deladriere** [comment]
Where do you set the CS pin to use ?

**Comment by agdl** [comment]
Tested. LGTM

**Comment by deladriere** [comment]
@agdl How do you set the CS pin ?

**Comment by tuxedo0801** [comment]
@cmaglie 

> you must patient a bit, this is still experimental, we'll provide documentation once ready.

It has been merged into master about three weeks ago. So it's ready now?

Questions so far:

* How do I enable this feature? Just include the "SDU" lib in my sketch? 
* Is it possible to use alternative pins for accessing sd-card? If yes, how?
* Is it somehow foreseen to easily use an alternative "SDUBoot.ino" implementation? (like using an SPI flash chip instead of a SD card)

**Comment by tuxedo0801** [comment]
For anyone searching the same information: #220

**Pull request merged**